File: app_node/box.js
'use strict';
var cfg = require('./config.js');
var cisco = require('./cisco.js');

var box = module.exports = {};

box.getConf = function() {
    return cfg.bcontrol;
}
box.checkAccess = function(boxname,secret) {
    var b;
    for(var i=0; b = cfg.bcontrol[i]; i++) {
        if(b.name === boxname && b.secret === secret) {
            return true;
        }
    }
    return false;
};
box.getStatus = function(boxname,type,id,ready) {
    //console.log('* request to get box='+boxname+' type='+type+' id='+id);
    var b;
    var m;
    for(var i=0; b = cfg.bcontrol[i]; i++) {
        if(b.name === boxname) {
            if(ready) {
                cfg.bcontrol[i].ready += 1;
                if(cfg.bcontrol[i].ready > 99999) {
                    cfg.bcontrol[i].ready = 0;
                }
            }
            for(var j=0; m = b.boxes[j]; j++) {
                if(m.type === type && m.id === parseInt(id)) {
                    console.log('* request box '+b.name+'/'+m.type+'/'+m.id+' reply = '+m.status);
                    return m.status;
                    //return cfg.bcontrol[i].boxes[j].status;
                }
            }
        }
    }
    return 0;
};
box.putOnFunc = function(boxname,type,id,fname) {
    var i = 0;
    while(cfg.bcontrol[i]) {
        var b = cfg.bcontrol[i];
        if(b.name === boxname) {
            var j = 0;
            while(b.boxes[j]) {
                var m = b.boxes[j];
                if(m.type === type && m.id === parseInt(id)) {
                    cfg.bcontrol[i].boxes[j].onfunc = fname;
                    return;
                }
                j++;
            }
        }
        i++;
    }
};
box.putOffFunc = function(boxname,type,id,fname) {
    var i = 0;
    while(cfg.bcontrol[i]) {
        var b = cfg.bcontrol[i];
        if(b.name === boxname) {
            var j = 0;
            while(b.boxes[j]) {
                var m = b.boxes[j];
                if(m.type === type && m.id === parseInt(id)) {
                    cfg.bcontrol[i].boxes[j].offfunc = fname;
                    return;
                }
                j++;
            }
        }
        i++;
    }
};
box.putRFunc = function(boxname,type,id,fname) {
    var i = 0;
    while(cfg.bcontrol[i]) {
        var b = cfg.bcontrol[i];
        if(b.name === boxname) {
            var j = 0;
            while(b.boxes[j]) {
                var m = b.boxes[j];
                if(m.type === type && m.id === parseInt(id)) {
                    cfg.bcontrol[i].boxes[j].rfunc = fname;
                    return;
                }
                j++;
            }
        }
        i++;
    }
};
box.putStatus = function(boxname,type,id,val,ready) {
    console.log('* request to change box='+boxname+' type='+type+' id='+id+' value='+val);
    var i = 0;
    while(cfg.bcontrol[i]) {
        var b = cfg.bcontrol[i];
        if(b.name === boxname) {
            if(ready) {
                cfg.bcontrol[i].ready += 1;
                if(cfg.bcontrol[i].ready > 99999) {
                    cfg.bcontrol[i].ready = 0;
                }
            }
            var j = 0;
            while(b.boxes[j]) {
                var m = b.boxes[j];
                if(m.type == type && m.id == id) {
                    //console.log('*** '+b.name+'/'+m.type+'/'+m.id+' changed to '+parseInt(val));
                    if(val != undefined) {
                        cfg.bcontrol[i].boxes[j].status = parseInt(val);
                    }
                    return;
                }
                j++;
            }
        }
        i++;
    }
};
module.exports = box;

File: app_python/box.py
import json
from app_python import app

boxcontrol = [
    {
        'name': 'b1',
        'secret': 'cisco123',
        'ready': 0,
        'boxes': [
            {
                'type': 'switch',
                'id': 0,
                'status': 0,
                'descr': 'red switch',
                'onfunc': 'anone',
                'offfunc': 'anone'
            },
            {
                'type': 'button',
                'id': 0,
                'status': 0,
                'descr': 'big red button',
                'onfunc': 'anone',
                'offfunc': 'anone'
            },
            {
                'type': 'usonic',
                'id': 0,
                'status': 0,
                'descr': 'ultrasonic sensor',
                'onfunc': 'anone',
                'offfunc': 'anone'
            },
            {
                'type': 'led',
                'id': 0,
                'status': 0,
                'descr': 'red light',
                'rfunc': 'rnone'
            },
            {
                'type': 'led',
                'id': 1,
                'status': 0,
                'descr': 'green light',
                'rfunc': 'rnone'
            }
        ]
    },
    {
        'name': 'b2',
        'secret': 'cisco123',
        'ready': 0,
        'boxes': [
            {
                'type': 'switch',
                'id': 0,
                'status': 0,
                'descr': 'blue switch',
                'onfunc': 'anone',
                'offfunc': 'anone'
            },
            {
                'type': 'button',
                'id': 0,
                'status': 0,
                'descr': 'big red button',
                'onfunc': 'anone',
                'offfunc': 'anone'
            },
            {
                'type': 'usonic',
                'id': 0,
                'status': 0,
                'descr': 'ultrasonic sensor',
                'onfunc': 'anone',
                'offfunc': 'anone'
            },
            {
                'type': 'led',
                'id': 0,
                'status': 0,
                'descr': 'red light',
                'rfunc': 'rnone'
            },
            {
                'type': 'led',
                'id': 1,
                'status': 0,
                'descr': 'green light',
                'rfunc': 'rnone'
            }
        ]
    }
]

def checkAccess(boxname,secret):
    for b in boxcontrol:
        if b['name'] == boxname and b['secret'] == secret:
            return True
    return False

def putStatus(boxname, t, i, v, r):
    for b in boxcontrol:
        #print('compare boxname',b['name'],boxname)
        if b['name'] == boxname:
            #print('boxnames equal',b['name'],boxname)
            if r:
                b['ready'] += 1
                if b['ready'] > 99999:
                    b['ready'] = 0;
            for m in b['boxes']:
                #print('compare type',m['type'],t)
                #print('compare id',m['id'],i)
                if m['type'] == t and int(m['id']) == int(i):
                    #print('bingo')
                    m['status'] = int(v)
                    return
            return
    return

def getStatus(boxname, t, i, r):
    for b in boxcontrol:
        if b['name'] == boxname:
            if r:
                b['ready'] += 1
                if b['ready'] > 99999:
                    b['ready'] = 0;
            for m in b['boxes']:
                if m['type'] == t and int(m['id']) == int(i):
                    return int(m['status'])
    return 0

def getConf():
    return json.dumps(boxcontrol)

def putOnFunc(boxname,t,i,fname):
    for b in boxcontrol:
        if b['name'] == boxname:
            for m in b['boxes']:
                if m['type'] == t and int(m['id']) == int(i):
                    m['onfunc'] = fname
                    return
    return

def putOffFunc(boxname,t,i,fname):
    for b in boxcontrol:
        if b['name'] == boxname:
            for m in b['boxes']:
                if m['type'] == t and int(m['id']) == int(i):
                    m['offfunc'] = fname
                    return
    return

def putRFunc(boxname,t,i,fname):
    for b in boxcontrol:
        if b['name'] == boxname:
            for m in b['boxes']:
                if m['type'] == t and int(m['id']) == int(i):
                    m['rfunc'] = fname
                    return
    return

if __name__ == '__main__':
  app.run()
